rfcs/rdm-0045-versioning-support.md

```diff
@@ -0,0 +1,530 @@
+
+
+#### Reindexing / denormalization
+
+- Reindex siblings.
+
+
+### Service layer
+
+#### Create a new record
+
+```python
+service.create(id_, identity, data, links_config=...)
+```
+
+#### Read a record
+
+```python
+service.read(id_, identity, data, links_config=...)
+service.read_draft(id_, identity, data, links_config=...)
+```
+
+#### Edit a record
+
+```python
+service.edit(id_, identity, data, links_config=...)
+```
+
+#### Create a new version
+
+```python
+service.new_version(id_, identity, data, links_config=...)
+```
+
+#### Updating a draft
+
+The update of a draft is unchanged, except that you can no longer change the ownership and permissions via the draft (which anyway was not possible in the February release).
+
+**Visibility and embargo information**
+
+Specifically, you will still update the record/files restrictions and embargo date/reason via the is method. The only difference is that it requires manage permissions to change this information, where as the metadata can be changed with edit permission.
+
+#### Updating ownership and permissions
+
+We propose adding a new method specifically update the permissions:
+
+```python
+service.update_permissions(id_, identity, data, links_config=...)
+service.update_owners(id_, identity, data, links_config=...)
+```
+
+- Resolve the id
+    - Should transparently handle the concept pid / pid issue (explicit - always use concept - not logical for APIs)
+- Check revision id
+- Check permissions
+- Load the data
+- Run components
+- Commit concept record
+- Index
+    - Reindex all records and drafts (bulk index)
+
+#### Search versions
+
+```python
+service.search_versions()
+```
+
+#### Reading a record/draft (dereferencing)
+
+(not needed? managed through record service)
+
+#### The service schema
+
+- A new schema for dumping permissions.
+
+### Presentation layer
+
+#### REST API
+
+The existing REST API, already define endpoints for the main operations discussed in the service:
+
+- ``GET /records`` - Search records.
+- ``GET /user/records`` - Search drafts.
+- ``POST /records`` - Create a new draft.
+- ``POST /records/:id/draft`` - Edit a record.
+- ``PUT /records/:id/draft`` - Update draft.
+- ``POST /records/:id/actions/publish`` - Publish a record.
+- ``GET /record/:id/versions`` - Search versions for a record.
+- ``POST /record/:id/versions`` - Create a new version.
+
+We will add a new endpoint for updating the permissions:
+
+- ``PUT /records/:id/access`` - Update permissions.
+
+We use ``PUT`` instead of e.g. ``PATCH /records/:id`` because we want to make use of the concept records version counter in the ETag for optimistic concurrency control, and we want to model the endpoint as a new REST resource. Some notes:
+
+- Identifier: We're using the ``:id`` of single version records instead of the concept record. This is to make it easier for API developers to deal only with a single identifier for a version, and to not confusing by using a different id.
```

Comment:
```suggestion
- Identifier: We're using the ``:id`` of single version records instead of the concept record. This is to make it easier for API developers: they only have to deal with a single identifier (the one for a record version), they don't have to make an additional API call, and they don't need to track another identifier.
```